relay-server/src/services/buffer/mod.rs

```diff
@@ -443,6 +410,111 @@ impl EnvelopeBufferService {
+        services: &Services,
+        buffer: &mut PolymorphicEnvelopeBuffer,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<(), EnvelopeBufferError> {
+        let own_key = project_key_pair.own_key;
+        let own_project = services.project_cache_handle.get(own_key);
+        // We try to load the project state and bail in case it's pending.
+        let own_project_info = match own_project.state() {
+            ProjectState::Enabled(info) => Some(info),
+            ProjectState::Disabled => None,
+            ProjectState::Pending => {
+                buffer.mark_ready(&own_key, false);
+                relay_statsd::metric!(
+                    counter(RelayCounters::BufferProjectPending) += 1,
+                    partition_id = &partition_tag
+                );
+
+                return Ok(());
+            }
+        };
+
+        let sampling_key = project_key_pair.sampling_key;
+        let mut sampling_project = None;
+        // If the keys are not the same, it means we have a different root project for this
+        // envelope, and we need to fetch the project config of the root project.
+        if project_key_pair.has_distinct_sampling_key() {
+            sampling_project = Some(services.project_cache_handle.get(sampling_key));
+        }
+
+        // We try to load the project state and bail in case it's pending.
+        let sampling_project_info =
+            if let Some(sampling_project_state) = sampling_project.as_ref().map(|p| p.state()) {
+                match sampling_project_state {
+                    ProjectState::Enabled(info) => Some(info),
+                    ProjectState::Disabled => None,
+                    ProjectState::Pending => {
+                        buffer.mark_ready(&sampling_key, false);
+                        relay_statsd::metric!(
+                            counter(RelayCounters::BufferProjectPending) += 1,
+                            partition_id = &partition_tag
+                        );
+
+                        return Ok(());
+                    }
+                }
+            } else {
+                None
+            };
+
+        relay_log::trace!("EnvelopeBufferService: popping envelope");
+
+        // If we arrived here, know that both projects are available, so we pop the envelope.
+        let envelope = buffer
+            .pop()
+            .await?
+            .expect("Element disappeared despite exclusive excess");
+
+        // If the project state is disabled, we want to drop the envelope and early return.
+        let Some(own_project_info) = own_project_info else {
+            let mut managed_envelope = ManagedEnvelope::new(
+                envelope,
+                services.outcome_aggregator.clone(),
+                services.test_store.clone(),
+                ProcessingGroup::Ungrouped,
+            );
+            managed_envelope.reject(Outcome::Invalid(DiscardReason::ProjectId));
+
+            return Ok(());
```

Comment:
Could this be an early return on line 424? Then `own_project_info` could be non-optional.